### DISCUSSION TITLE
Seeking clarification on Lab 3 single-cycle multipliers and enhancement

### DISCUSSION THREAD

**Discussion by hsienrong** [NUS-CG3207/labs#18] [Lab 3] [answered]
Hi TA's! In the Lab 3 manual, it is stated:

"DO NOT implement a single cycle multiplier - FPGAs have built in multipliers/DSP units, which are inferred when you use '*' operator. This is much more efficient than any array adder based multipliers you can implement, but we don't want to be using it in CG3207."

We would like to clarify on this statement. Does this mean:  
(A) We _can_ implement a single-cycle array adder based multiplier (just as shown in the attached image from the lecture slides), as long as we do not use the * operator,  
(B) We are not allowed to implement any kind of single-cycle multipliers for this lab?

![image](https://github.com/user-attachments/assets/a86e6346-c149-46e9-91bd-b329452f98ca)

**Comment by r-panicker** [comment, marked as answer]
Agree the phrasing could have been better.

There's not much point in coding an array based multiplier (other than as a learning exercise maybe), as it'll be much slower than the built-in multiplier. The latter will be inferred/used when you use the '*' operator.
And since you can't really improve single-cycle multiplication, the idea/suggestion was to try and improve the sequential multipler instead. 

However, given the confusion, if you have already implemented an array multiplier and wants to use that as the enhancement, you can.

**Reply by onx001** [reply]
Hi, then can I confirm if it will be eligible for the 5 marks as it is strictly speaking near the best performance enhancement we can get without using the inbuilt multiplier block? Thank you!

**Reply by r-panicker** [reply]
Didn't reply here earlier as I explained it to your teammate. The idea isn't to come up with the best, but to explore some tradeoffs. There is no 'best'; there are only good aka [Pareto-optimal](https://en.wikipedia.org/wiki/Pareto_efficiency) designs :)
